apis/supernova/src/handler/hypercube-handler.js

```diff
@@ -70,10 +54,52 @@ class HyperCubeHandler extends DataPropertyHandler {
   }
 
   getDimensionLayouts() {
-    const hc = getHyperCube(this.layout, this.path);
+    const hc = hcHelper.getHyperCube(this.layout, this.path);
     return hc ? hc.qDimensionInfo : [];
   }
 
+  addDimension(dimension, alternative, idx) {
+    const dim = initializeField(dimension);
+
+    if (hcHelper.isDimensionAlternative(this, dim, alternative)) {
+      return hcHelper.addAlternativeDimension(this, dim, idx);
+    }
+
+    return addMainDimension(this, dim, idx);
+  }
+
+  async addDimensions(dimensions, alternative = false) {
+    const existingDimensions = this.getDimensions();
+    const addedDimensions = [];
+    let addedActive = 0;
+    await Promise.all(
+      dimensions.map(async (dimension) => {
+        if (hcHelper.isTotalDimensionsExceeded(this, existingDimensions)) {
+          return addedDimensions;
+        }
+
+        const dim = initializeField(dimension);
+
+        if (hcHelper.isDimensionAlternative(this, alternative)) {
+          const altDim = await hcHelper.addAlternativeDimension(this, dim);
+          addedDimensions.push(altDim);
+        } else if (existingDimensions.length < this.maxDimensions()) {
+          await hcHelper.addActiveDimension(this, dim, existingDimensions, addedDimensions, addedActive);
+          addedActive++;
+        }
+        return addedDimensions;
+      })
```

Comment:
I do expect addDimensions to add thing in same order as dimensions array.
this looks to add them in the order of what is done first.

Reply:
Good catch. But I do think it does add them in the correct order since the "awaited" call is not actually an async function and gets resolved directly.

Either way, the calls inside `addDimensions` does have to be async. But the method itself might still need to be, since it's possible called as an async function but some code that consumes it.

Reply:
Turns out `addActiveDimension` is making something that looks like an async call. So you're right, there is a risk of a race condition here.

I would then think `for await (const value of object) {}` is the way to go. Instead of using `dimensions.map`.

Reply:
@T-Wizard @cbt1 I ended up using `await array.reduce()` instead. Apparently for await requires regenerator-runtime, which is too heavyweight for it to allow as Eslint error says. The best replacement which follows the same concept is reduce, it waits for each previous round promise to be resolved and then it goes to the next one.